stan/math/prim/mat/err/check_matching_dims.hpp

```diff
@@ -47,6 +46,49 @@ namespace stan {
                        "columns of ", name2, y2.cols());
     }
 
+    /**
+     * Check if the two matrices are of the same size.
+     *
+     * This function checks the runtime sizes and can also check the static
+     * sizes as well. For example, a 4x1 matrix is not the same as a vector
+     * with 4 elements.
+     *
+     * @tparam check_compile Whether to check the static sizes 
+     * @tparam T1 Scalar type of the first matrix
+     * @tparam T2 Scalar type of the second matrix
+     * @tparam R1 Rows specified at compile time of the first matrix
+     * @tparam C1 Columns specified at compile time of the first matrix
+     * @tparam R2 Rows specified at compile time of the second matrix
+     * @tparam C2 Columns specified at compile time of the second matrix
+     *
+     * @param function Function name (for error messages)
+     * @param name1 Variable name for the first matrix (for error messages)
+     * @param y1 First matrix
+     * @param name2 Variable name for the second matrix (for error messages)
+     * @param y2 Second matrix
+     *
+     * @throw <code>std::invalid_argument</code> if the 
+     * dimensions of the matrices do not match
+     */
+    template <bool check_compile, typename T1, typename T2,
+              int R1, int C1, int R2, int C2>
+    inline void check_matching_dims(const std::string& function,
+                                    const std::string& name1,
+                                    const Eigen::Matrix<T1, R1, C1>& y1,
+                                    const std::string& name2,
+                                    const Eigen::Matrix<T2, R2, C2>& y2) {
+      if (check_compile) {
```

Comment:
You should combine these to be 
```
if (check_compile && (R1 != R2 || C1 != C2)) { 
  ...
}
```
Anything that saves indentation is a good refactoring.